### ISSUE TITLE
Cucumber spring does not support multi-threaded scenarios

### ISSUE THREAD

**Issue by ishmeister** [cucumber/cucumber-jvm#1106]
## Summary

I've set up a prototype project using cucumber-jvm-parallel-plugin and maven-surefire-plugin. I've set surefire to run parallel test classes within the same JVM (threaded).

This set up is supposed to create a JUnit test class per feature (which it does) and run them in parallel threads (which it also does). The alternative to threaded mode is to run a fork of the JVM with each feature, which obviously requires more system resources but there is complete isolation between each test. The bug I describe here does not apply in the latter case.

## Expected Behavior

Before adding the cucumber-spring dependency to maven, I get (as I expect) one instance of the StepDef class per scenario even when the scenarios are multi-threaded:

~~~
10:46:05.924 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - before scenario: Search for Badger
10:46:05.929 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - i am object: 2132331383
10:46:13.993 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - after scenario: Search for Badger
10:46:13.993 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - i am object: 2132331383

10:46:05.924 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - before scenario: Search for Llama
10:46:05.929 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - i am object: 1544373948
10:46:12.365 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - after scenario: Search for Llama
10:46:12.365 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - i am object: 1544373948
~~~

There is no conflict between scenarios. Each scenario runs in a separate browser window.

## Current Behavior

However, when I add the cucumber-spring dependency (which changes how StepDef classes are instantiated) this isolation between scenarios breaks down:

The result is that the two scenarios are now sharing an StepDef instance and therefore the WebDriver, which is causing conflicts.

~~~
10:56:49.741 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - before scenario: Search for Badger
10:56:49.745 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - i am object: 616523439
10:56:56.165 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - after scenario: Search for Badger
10:56:56.166 [pool-1-thread-1] INFO  c.m.test.steps.WikipediaSteps - i am object: 874567236

10:56:49.741 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - before scenario: Search for Llama
10:56:49.745 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - i am object: 1829050026
10:56:52.869 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - after scenario: Search for Llama
10:56:52.869 [pool-1-thread-2] INFO  c.m.test.steps.WikipediaSteps - i am object: 616523439
~~~

There is a conflict between scenarios. "after scenario: Search for Llama" prematurely ends "scenario: Search for Badger".

## Possible Solution
StepDef spring beans should have both scenario and thread scope.

## Steps to Reproduce (for bugs)

```
Feature: Search on Wikipedia for Badgers
  Scenario: Search for Badger
    Given I am on the Wikipedia homepage
    When I search for "Badger"
    Then I am on the page with title "Badger"
```

```
Feature: Search on Wikipedia for Llamas
  Scenario: Search for Llama
    Given I am on the Wikipedia homepage
    When I search for "Llama"
    Then I am on the page with title "Llama"
```

```
public class WikipediaSteps extends FluentCucumberTest {

    private static final Logger logger = LoggerFactory.getLogger(WikipediaSteps.class);

    @Given("^I am on the Wikipedia homepage")
    public void I_am_on_the_homepage() throws Throwable {
        goTo("https://en.wikipedia.org");
    }

    @When("^I search for \"(.+)\"$")
    public void I_search_for(String searchText) throws Throwable {
        $("#searchInput").fill().with(searchText);
        $("#searchButton").click();
    }

    @Then("^I am on the page with title \"(.+)\"")
    public void i_am_on_the_page_with_title(String expectedHeading) throws Throwable {
        String heading = $("#firstHeading").textContent();
        assertEquals(expectedHeading, heading);
    }

    @Before
    public void before(Scenario scenario) {
        logger.info("before scenario: {}", scenario.getName());
        logger.info("i am object: {}", System.identityHashCode(this));
        super.before(scenario);
    }

    @After
    public void after(Scenario scenario) {
        logger.info("after scenario: {}", scenario.getName());
        logger.info("i am object: {}", System.identityHashCode(this));
        super.after(scenario);
    }
}
```

```
<plugin>
    <groupId>com.github.temyers</groupId>
    <artifactId>cucumber-jvm-parallel-plugin</artifactId>
    <version>2.2.0</version>
    <executions>
        <execution>
            <id>generateRunners</id>
            <phase>validate</phase>
            <goals>
                <goal>generateRunners</goal>
            </goals>
            <configuration>
                <glue>com.mycompany.test.steps</glue>
                <featuresDirectory>src/test/resources/cucumber</featuresDirectory>
                <cucumberOutputDir>target/cucumber-parallel</cucumberOutputDir>
                <format>json,html</format>
                <tags>"~@ignored"</tags>
            </configuration>
        </execution>
    </executions>
</plugin>
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-surefire-plugin</artifactId>
    <version>2.19</version>
    <configuration>
        <threadCount>2</threadCount>
        <parallel>classes</parallel>
        <includes>
            <include>**/*IT.class</include>
        </includes>
    </configuration>
</plugin>
```
## Context & Motivation

Running features in parallel without the requirement to fork the JVM.
Faster test execution and more efficient use of build server resources.

## Your Environment

* Version used: 
        <selenium.version>3.3.1</selenium.version>
        <cucumber.version>1.2.5</cucumber.version>
        <fluentlenium.version>3.1.1</fluentlenium.version>
        <spring.version>4.3.7.RELEASE</spring.version>
* Operating System and version: Windows 10
* Link to your project:

**Comment by ishmeister** [comment]
I have fixed this issue and created a pull request.

**Comment by mpkorstanje** [comment]
Cheers! This fix will be available once 2.0.0 is done.
